ddtrace/profiling/collector/stack.pyx

```diff
@@ -281,6 +302,58 @@ cdef stack_collect(ignore_profiler, thread_time, max_nframes, interval, wall_tim
     return stack_events, exc_events
 
 
+@attr.s(slots=True, eq=False)
+class _ThreadSpanLinks(object):
+
+    _thread_id_to_spans = attr.ib(factory=lambda: collections.defaultdict(weakref.WeakSet), repr=False, init=False)
+
+    # We do not use a lock because:
+    # - When adding new items, it's not possible for a same thread to call `link_span` at different time
+    #   Since the WeakSet are per-thread, there's no chance of creating 2 WeakSet for the same thread
+    # - When reading the span WeakSet for a thread, the set is copied which means that if it was ever mutated during
+    #   our reading, we wouldn't care.
+    #   In practice, here, it won't even mutate during the reading since this only used by the stack collector which
+    #   already owns the GIL.
+
+    def link_span(self, span):
+        """Link a span to its running environment.
+
+        Track threads, tasks, etc.
+        """
+        self._thread_id_to_spans[_thread_get_ident()].add(span)
+
+    def clear_threads(self, existing_thread_ids):
+        """Clear the stored list of threads based on the list of existing thread ids.
+
+        If any thread that is part of this list was stored, its data will be deleted.
+
+        :param existing_thread_ids: A set of thread ids to keep.
+        """
+        for thread_id in list(self._thread_id_to_spans.keys()):
+            if thread_id not in existing_thread_ids:
+                del self._thread_id_to_spans[thread_id]
+
+    def get_active_child_spans_from_thread_id(self, thread_id):
```

Comment:
These aren't necessarily just child spans being returned (if there is a root span that isn't finished it will be returned) so the function name doesn't make sense. "Leaf" spans would probably make more sense here.

Reply:
Good suggestion. Changed.